### ISSUE TITLE
Debugging in Cranelift

### ISSUE THREAD

**Issue by lachlansneff** [CraneStation/cranelift#426]
It's become clear that Cranelift needs the facilities to debug the compiled code.

I propose implementing a trait `DebugSink` that looks something like this:

```rust
trait DebugSink {
  fn insert_inst(&mut self, inst: InstructionData, source_loc: SourceLoc, code_offset: CodeOffset);
  fn insert_func(&mut self, name: String, source_loc: SourceLoc, code_offset: CodeOffset);
}
```

Once `insert_func` is called, all instructions afterwards (until another function is inserted) belong to that function.

It could also take on the form of two traits:

```rust
trait DebugSinkFunc {
  fn insert_inst(&mut self, inst: InstructionData, source_loc: SourceLoc, code_offset: CodeOffset);
}

trait DebugSink {
  fn insert_func(&mut self, name: String, source_loc: SourceLoc, code_offset: CodeOffset) -> &mut DebugSinkFunc
}
```

**Comment by sunfishcode** [comment]
I think it's clear we need something like this. What follows are some misc. thoughts about the interface.

It feels like this might be something we could merge into the `CodeSink` trait. In addition to reducing the number of sink traits we have, that would also avoid the need to pass in the `CodeOffset`.

Is there a need for passing in the whole `InstructionData`?

One use case I'm envisioning for this is https://github.com/CraneStation/cranelift/issues/350. For that, we just need the `SourceLoc`, which in this context could be used to whole the line/column/file/etc. information of an instruction.

**Comment by lachlansneff** [comment]
I agree with you on merging debugging into the `CodeSink` trait. As for `InstructionData`, I was just trying to find some way of supplying the current instruction.

**Comment by philipc** [comment]
If this is merged into `CodeSink`, then how would `MemoryCodeSink` implement it? Would it still need a `DebugSink` to delegate the methods to (similar to `RelocSink` and `TrapSink`)? Doing that would add a lot of virtual calls (the opposite of why `MemoryCodeSink` exists).

Or would `MemoryCodeSink` do nothing for debug, and so you either can't use it if you want debug, or you need to wrap it in something else that handles the debug, eg:
```rustc
struct DwarfCodeSink<`a> {
    mem: MemoryCodeSink<'a>,
    // Fields for the DWARF generation
    ...
}

impl<'a> CodeSink for DwarfCodeSink<'a> {
   // Delegate to self.mem for everything except the debug methods
   ...
}
```

**Comment by sunfishcode** [comment]
For most DWARF sections, would it work to generalize `begin_rodata()` into a more general mechanism to "switch sections"?

For debug_line sections, I have two possible ideas so far.

For one, I wonder if we could add a function to `CodeEmitter` which would have roughly the interface of the [`.loc`](https://sourceware.org/binutils/docs/as/Loc.html#Loc) directive. Something like this in `CodeEmitter`:
```rust
    fn debug_loc(file: SomeFileID, lineno: u32, column: u32, options: SomeOptionsSet)
```
and then we call the actual debug_line-emitting code from within the trait, rather than outside the trait.

For another, I think we have enough information to be able to emit a debug_line section in a separate pass from emitting the instructions, so then we could just use whatever mechanism we added for switching sections, and the normal `CodeEmitter` interface.

Thoughts?

**Comment by philipc** [comment]
I assume you mean `CodeSink` rather than `CodeEmitter`.

> For most DWARF sections, would it work to generalize begin_rodata() into a more general mechanism to "switch sections"?

Probably not. The sections will be built up by gimli, and you'll just get `Vec`s with the complete section data out of it (plus relocations). The problem we need to solve is how to give gimli the info it needs. A lot of the `.debug_info` generation needs to be driven at a higher level  (eg in rustc_codegen_cranelift) since it needs to know about things like types. `.debug_info` then contains references to other sections, many of which need to be built up during code generation:
- .eh_frame - needs to know register stores/loads (or more? not sure)
- .debug_line - needs to know source locations for each instruction
- .debug_loclists - needs to know variable locations at each instruction address
- .debug_rnglists - needs to know which address ranges belong to each inlined function

Now if this info is all available without needing to hook into `CodeSink`, then doing these as separate passes sounds great. We can either pass this info directly to gimli, or do it via a ` DebugSink` trait. I don't think this would use `CodeSink` or a "switch sections" mechanism. gimli may be able to output to a `CodeSink`, but I don't see any benefit compared to gimli writing to a `Vec`.

> we call the actual debug_line-emitting code from within the trait

This seems like you think the debug_line-emitting code will be low level `put1`/`put2` etc? Instead I think this will need to call into gimli, which will have some associated state. Where would that state be kept? Will all implementations of `CodeSink` need to include some state for DWARF generation? This is related to my question of what would `MemoryCodeSink` do? If DWARF generation is separate from `CodeSink` though, then we avoid this problem.

**Comment by philipc** [comment]
I think I'll be able to iterate through the instruction offsets and srclocs using `Function::inst_offsets()` and `Function::srclocs`. That should be enough to hack something together for `.debug_line` without touching `CodeSink`. I first need to do the `.debug_line` support in gimli though.
